pkg/controller/kogitoinfra/mongodb.go

```diff
@@ -0,0 +1,366 @@
+	return []framework.WatchedObjects{
+		{
+			GroupVersion: mongodb.SchemeGroupVersion,
+			AddToScheme:  mongodb.SchemeBuilder.AddToScheme,
+			Objects:      []runtime.Object{&mongodb.MongoDB{}},
+		},
+		{
+			Objects: []runtime.Object{&corev1.Secret{}},
+		},
+	}
+}
+
+// Reconcile reconcile Kogito infra object
+func (i *mongoDBInfraReconciler) Reconcile() (requeue bool, resultErr error) {
+	var mongoDBInstance *mongodb.MongoDB
+
+	if !infrastructure.IsMongoDBAvailable(i.client) {
+		return false, errorForResourceAPINotFound(&i.instance.Spec.Resource)
+	}
+
+	// Step 1: check whether user has provided custom mongoDB instance reference
+	mongoDBNamespace := i.instance.Spec.Resource.Namespace
+	mongoDBName := i.instance.Spec.Resource.Name
+	if len(mongoDBName) > 0 {
+		log.Debugf("Custom mongodb instance reference is provided")
+		if len(mongoDBNamespace) == 0 {
+			mongoDBNamespace = i.instance.Namespace
+			log.Debugf("Namespace is not provided for custom resource, taking instance namespace(%s) as default", mongoDBNamespace)
+		}
+	} else {
+		mongoDBName = infrastructure.MongoDBInstanceName
+		mongoDBNamespace = i.instance.Namespace
+		// create/refer kogito-mongodb instance
+		log.Debugf("Custom mongodb instance reference is not provided")
+		// Verify MongoDB Operator (it's installation is required in the same namespace, that's why we do this check as well)
+		if mongoDBAvailable, err := infrastructure.IsMongoDBOperatorAvailable(i.client, i.instance.Namespace); err != nil {
+			return false, err
+		} else if !mongoDBAvailable {
+			return false, errorForResourceAPINotFound(&i.instance.Spec.Resource)
+		}
+	}
+
+	if mongoDBInstance, resultErr = i.loadDeployedMongoDBInstance(mongoDBName, mongoDBNamespace); resultErr != nil {
+		return false, resultErr
+	} else if mongoDBInstance == nil {
+		return false, errorForResourceNotFound("MongoDB", i.instance.Spec.Resource.Name, mongoDBNamespace)
+	}
+
+	log.Debugf("Got MongoDB instance %v", mongoDBInstance)
+	if mongoDBInstance.Status.Phase != mongodb.Running {
+		return false, errorForResourceNotReadyError(fmt.Errorf("mongoDB instance %s not ready. Waiting for Status.Phase == Running", mongoDBInstance.Name))
+	}
+	log.Info("MongoDB instance is running")
+	if resultErr = i.updateMongoDBAppPropsInStatus(mongoDBInstance); resultErr != nil {
+		return false, nil
+	}
+	if resultErr = i.updateMongoDBEnvVarsInStatus(mongoDBInstance); resultErr != nil {
+		return false, nil
+	}
+	return false, resultErr
+}
+
+// MongoDBCredential holds the credentials information of a user into an mongoDB server
+type MongoDBCredential struct {
+	AuthDatabase string `yaml:"auth-database"`
+	Username     string `yaml:"username"`
+	Password     string `yaml:"password"`
+	Database     string `yaml:"database"`
+}
+
+// fillMongoDBCredential fills the credential of the MongoDB server deployed with the Kogito Operator.
+// Prefilled values will be taken into account and corrected if needed.
```

Comment:
Is it a good approach to correct provided values? Shouldn't we rather throw an error complaining about wrong values provided?

Reply:
This is something I put in place when i was also creating an instance when none was provided.
Seems good not to have default values and raise an error indeed.
So we should consider that the MongoDB will always need Name, AuthDB and Database in the KogitoInfra Config part.
@ricardozanini wdyt ?

Reply:
We need to implement Admission controllers, actually. So we could validate the CR before it's created. 

I think raising an error is a good approach, but best if we can have a named error, so users would clearly see what's wrong via `status` attribute of the `KogitoInfra` CR. And also the controller won't reconcile anymore.

Adding defaults, in this case, **could** lead to hard to catch errors during runtime if the connection information was incorrectly set by the operator just because we defined default values.

Reply:
oki. gonna update that